polynote-server/src/main/scala/polynote/server/SharedNotebook.scala

```diff
@@ -324,8 +328,50 @@ class IOSharedNotebook(
 
     override def info: IO[Option[KernelInfo]] = ifKernelStarted(_.info, None)
 
-    override def getHandleData(handleType: HandleType, handle: Int, count: Int): IO[List[ByteVector32]] =
-      ensureKernel().flatMap(_.getHandleData(handleType, handle, count))
+    private val streams = new mutable.HashMap[Int, Iterator[ByteVector32]]
+
+    private class HandleIterator(handleId: Int, iter: Iterator[ByteVector32]) extends Iterator[ByteVector32] {
+      override def hasNext: Boolean = {
+        val hasNext = iter.hasNext
+        if (!hasNext) {
+          streams.remove(handleId)  // release this iterator for GC to make underlying collection available for GC
```

Comment:
tbh not 100% clear on WeakHashMaps but would this be a place where using one would help?

Reply:
Hmm, well we *don't* want it to get GC'ed until it's done. There is probably more work needed here to make things release nicely though.